src/controller/python/chip/ChipStack.py

```diff
@@ -187,23 +183,28 @@ def GetStorageManager(self):
     def enableServerInteractions(self):
         return self._enableServerInteractions
 
+    def RegisterSubscription(self, subscription: ClusterAttribute.SubscriptionTransaction):
+        self._subscriptions[subscription.subscriptionId] = subscription
+
+    def UnregisterSubscription(self, subscription: ClusterAttribute.SubscriptionTransaction):
+        del self._subscriptions[subscription.subscriptionId]
+
     def Shutdown(self):
-        #
+
+        # Shutdown all subscriptions before shutting down the stack.
+        for subscription in tuple(self._subscriptions.values()):
```

Comment:
I don't think this is required?

```suggestion
        for subscription in self._subscriptions.values():
```

Reply:
This is required because when the `Shutdown()` is called the subscription is removed from the _subscriptions dict. And Python does not allow dict modification during iteration.

Reply:
This just sounds like you are relying on tuple to create a local copy of things to iterate over. Would a more readable way be to just store it locally and then iterate through your copy instead?

Reply:
Afaik using tuple() to create a copy is actually quite commonly used. I didn't realize that a copy is needed here, but make sense. Maybe adding a comment would still be nice.

Reply:
OK, I'll add a comment what is happening here.